muacryptcc/plugin.py

```diff
@@ -6,41 +6,91 @@
 from hippiehug import Chain
 from claimchain import State, View
 from claimchain.crypto.params import LocalParams
-from claimchain.utils import pet2ascii
+from claimchain.utils import pet2ascii, ascii2pet
+from muacrypt.mime import parse_email_addr, get_target_emailadr
+from .filestore import FileStore
 
 hookimpl = pluggy.HookimplMarker("muacrypt")
 
 
 @hookimpl
 def instantiate_account(plugin_manager, basedir):
     cc_dir = os.path.join(basedir, "muacryptcc")
-    cc_manager = CCAccount(cc_dir)
+    store_dir = os.path.join(cc_dir, "store")
+    store = FileStore(store_dir)
+    cc_manager = CCAccount(cc_dir, store)
     plugin_manager.register(cc_manager)
 
 
-class CCAccount:
+class CCAccount(object):
     def __init__(self, accountdir, store=None):
         self.accountdir = accountdir
         if not os.path.exists(accountdir):
             os.makedirs(accountdir)
+        self.addr2root_hash = {}
+        self.addr2pk = {}
         self.store = store
         self.init_crypto_identity()
 
+    #
+    # muacrypt plugin hook implementations
+    #
+    @hookimpl
+    def process_incoming_gossip(self, addr2pagh, account_key, dec_msg):
+        root_hash = dec_msg["GossipClaims"]
+        store = FileStore(dec_msg["ChainStore"])
+        peers_chain = Chain(store, root_hash=ascii2pet(root_hash))
+        assert peers_chain
+        view = View(peers_chain)
+        peers_pk = view.params.dh.pk
+        assert peers_pk
+        sender = parse_email_addr(dec_msg["From"])
+        self.addr2root_hash[sender] = root_hash
+        self.addr2pk[sender] = peers_pk
+        import logging ; logging.debug(sender + " added")
+        recipients = get_target_emailadr(dec_msg)
+        # TODO: handle everyone.
+        for recipient in recipients:
+            value = self.read_claim_from(peers_chain, recipient)
+            if value:
+                # for now we can only read claims about ourselves...
+                # so if we get a value it must be our head.
+                assert value == self.head
+                assert 0
+
+    @hookimpl
+    def process_outgoing_before_encryption(self, account_key, msg):
+        recipients = get_target_emailadr(msg)
+        for recipient in recipients:
+            key = recipient
+            value = self.addr2root_hash[recipient]
+            peers_pk = self.addr2pk[recipient]
+            if peers_pk:
+                self.add_claim(claim=(key, value), access_pk=peers_pk)
+            else:
+                import logging ; logging.debug(recipient + " not found")
+        self.commit_to_chain()
+        msg["GossipClaims"] = pet2ascii(self.head)
```

Comment:
``head`` should not be a petlib crypto object, but probably either raw bytes or hexdigest. I think this is not the right conversion to do.

Reply:
ah... now i see where pet is coming from. It's raw bytes. So i guess some conversion to hex is what we want.